### PR TITLE
Build_system.prefix_rules

### PR DESCRIPTION
**Pull request by rgrinberg**
Add a way to implicitly add a build prefix to all rules added by the given
function.

I'm not sure if I'm changing the right `add_rule` here. Because the way I have it now, there's nothing that guarantees that the prefix itself doesn't introduce any targets outside of `dir`. So I can move these checks to `Super_context.add_rule`, though it will be a lot more ugly.

### CODE REVIEW

**Comment by diml** [comment]
We can check in `prefix_rules` the the prefix contains no targets. That will simplify things.

There is something to do about `on_load_dir` as well, i.e. it needs to capture the prefix and we need to restore it when we run the callbacks.

**Comment by rgrinberg** [comment]
> We can check in prefix_rules the the prefix contains no targets. That will simplify things.

OK. Will do.

> There is something to do about on_load_dir as well, i.e. it needs to capture the prefix and we need to restore it when we run the callbacks.

Done. Double check that I did this right though. The way I'm doing it, we end up verifying the same has no targets more than once which is perhaps a bit unsatisfying.

**Comment by rgrinberg** [comment]
Nevermind. My last commit moves the check so that it's only done once per prefix.

File: src/build_system.mli [hunk 1]
```diff
@@ -47,6 +47,8 @@ val set_rule_generators : t -> (dir:Path.t -> string list -> extra_sub_directori
 *)
 val add_rule : t -> Build_interpret.Rule.t -> unit
 
+val prefix_rules : t -> (unit, unit) Build.t -> f:(unit -> 'a) -> 'a
```
Comment (diml) [inline src/build_system.mli]: This function needs a comment

**Comment by diml** [comment]
Ok, code looks good

**Comment by rgrinberg** [comment]
@diml docs added.

**Pull request merged**